Refactor creation of static variables

mwarusz opened the pull request:
<!--
Please add a description of what is accomplished in the PR here at the top:
-->

<!--
Below are a few things we ask you or your reviewers to kindly check. 
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist
* [x] Documentation:
  * [x] User's Guide has been updated
  * [x] Developer's Guide has been updated
  * [x] Documentation has been [built locally](https://e3sm-project.github.io/Omega/develop/devGuide/BuildDocs.html) and changes look as expected
* [x] Testing
  * [x] Unit tests have passed. Please provide a relevant CDash build entry for verification.
<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

@philipwjones This is my proposal to refactor the way we handle creation of static instances of main Omega classes. To present this, I changed the `MachEnv` class. The basic idea is to disallow calling constructors directly and only allow creating new instances by static `create` methods. Afterwards, the instances can only be manipulated through a pointer.

Some notes:
- Most other classes that use this pattern won't need `create` method implemented using variadic templates, at least initially, since they have only one constructor.
- it is still possible to disallow calling `MachEnv::create` with just a communicator for non-default environments. I did not do it, beacause I wanted to present the main idea as clearly as possible.

philipwjones reviewed (approved):
Thanks @mwarusz - we did take this approach with some of the later infra code like MetaData and IOFields, though I think we're using shared_ptr which is a bit less strict. I think with these early classes, I had tried using regular pointers (instead of smart pointers) and had some issues with scoping.  So I'm fine with this plan and you can feel free to update code as needed.

mwarusz commented:
@philipwjones This is now ready. I have rebased, updated every class to use this pattern, and updated the docs. In addition, I made two unrelated, hopefully acceptable,  changes in `MachEnv`:
- renamed `getEnv` and `getDefaultEnv` to `get` and `getDefault`, respectively. This is for consistency with other classes.
- switched to using `LOG_ERROR` instead of `std::cerr` for errors, since logging is now available.

Test pass on Chrysalis, Frontier cpu and Perlmutter gpu. ( I was not able to run them on Frontier gpu after the module update).

philipwjones commented:
Thanks @mwarusz this looks good and the routine name changes are fine. I had to push one other change to replace some remaining std::cout calls in MachEnv that were preventing successful builds on Chrys/Intel and Frontier. But with that change, all ctests pass on Chrys/Intel and Frontier (cpu/gpu).
